src/Microsoft.Azure.WebJobs.Host/Bindings/QueueBindingRules.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Microsoft.Azure.WebJobs.Host.Queues;
+using Microsoft.Azure.WebJobs.Host.Storage;
+using Microsoft.Azure.WebJobs.Host.Storage.Queue;
+using Microsoft.WindowsAzure.Storage.Queue;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+    // Write up bindinging rules for [Queue] attribute. 
+    // This is fundemantentally an IAsyncCollector<IStorageQueueMessage>
+    internal class QueueBindingRules
+    {
+        private readonly IStorageAccountProvider _accountProvider;
+        private readonly IContextGetter<IMessageEnqueuedWatcher> _messageEnqueuedWatcherGetter;
+
+        public QueueBindingRules(IStorageAccountProvider accountProvider, IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter)
+        {
+            _accountProvider = accountProvider;
+            _messageEnqueuedWatcherGetter = messageEnqueuedWatcherGetter;
+        }
+
+        public IBindingProvider BuildQueueRules(
+            INameResolver nameResolver,
+            IConverterManager converterManager)
+        {
+            // IStorageQueueMessage is the core testing interface 
+            converterManager.AddConverter<byte[], IStorageQueueMessage, QueueAttribute>(ConvertByteArrayCloudQueueMessage);
+            converterManager.AddConverter<string, IStorageQueueMessage, QueueAttribute>(ConvertString2CloudQueueMessage);
+            converterManager.AddConverter<IStorageQueueMessage, string>(ConvertCloudQueueMessage2String);
+            converterManager.AddConverter<IStorageQueueMessage, byte[]>(ConvertCloudQueueMessage2ByteArray);
+
+            converterManager.AddConverter<CloudQueueMessage, IStorageQueueMessage>(ConvertReal2CloudQueueMessage);
+
+            var bf = new BindingFactory(nameResolver, converterManager);
+
+            var ruleQueueOutput = bf.BindToAsyncCollector<QueueAttribute, IStorageQueueMessage>(BuildFromQueueAttribute, ToParameterDescriptorForCollector, FixerUpper);
+            var ruleQueueClient = bf.BindToExactAsyncType<QueueAttribute, IStorageQueue>(BuildClientFromQueueAttributeAsync, ToParameterDescriptorForCollector, FixerUpper);
+            var ruleQueueClient2 = bf.BindToExactAsyncType<QueueAttribute, CloudQueue>(BuildRealClientFromQueueAttributeAsync, ToParameterDescriptorForCollector, FixerUpper);
+            var queueRules = new GenericCompositeBindingProvider<QueueAttribute>(
```

Comment:
rename queueRules to bindingProvider. I'd like us not to have name mismatches like this :)